### PR TITLE
Chapter 2 Solutions

### PR DESCRIPTION
**Pull request by Volence**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -832,7 +859,8 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate rotateLength l = drop rotateLength $ take (length l + rotateLength) (cycle l)
```
Comment (vrom911) [inline src/Chapter2.hs:863]: Unfortunately, `cycle` fails in runtime on empty lists :recycle: So you need to handle this case separately for this function to work properly.

Also, you should consider how to handle negative `rotateLength` as well 🙂

File: src/Chapter2.hs [hunk 2]
```diff
@@ -848,7 +876,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x:xs) = (rewind xs) ++ [x]
```
Comment (vrom911) [inline src/Chapter2.hs:881]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by vrom911 — APPROVED** [review]
Amazing! You have really impressive solutions with interesting ideas! Well done 🎉

**Comment by chshersh** [comment]
I'm going to close this PR because we can't merge the solutions to our repo :slightly_smiling_face:
Feel free to open PRs with solutions to other chapters!

**Pull request closed without merging**